neo/Persistence/LevelDB/DbSnapshot.cs

```diff
@@ -22,6 +22,7 @@ internal class DbSnapshot : Snapshot
         public override DataCache<UInt256, AssetState> Assets { get; }
         public override DataCache<UInt160, ContractState> Contracts { get; }
         public override DataCache<StorageKey, StorageItem> Storages { get; }
+        public override DataCache<MPTKey, MerklePatricia> MPTStorages { get; }
```

Comment:
The entire MerklePatricia tree cannot be serialized and deserialized each time a block completes that changes storage of one of the items in the tree. The implementation needs to store the actual entries of the Merkle Patricia tree in separate keys in the LevelDB database, so that the entire tree is not written into level db for each contract that has had any storage change each block. This change is possible and required to be able to adopt the use of the MPT.

Reply:
I should say though; great job in implementing the MPT! I look forward to seeing the changes implemented in the future to be able to efficiently write changes to MPT into the persistent database.

Reply:
> The entire MerklePatricia tree cannot be serialized and deserialized each time a block completes that changes storage of one of the items in the tree. The implementation needs to store the actual entries of the Merkle Patricia tree in separate keys in the LevelDB database, so that the entire tree is not written into level db for each contract that has had any storage change each block. This change is possible and required to be able to adopt the use of the MPT.

Yeah, actually @igormcoelho already mentioned that with me, at that time I thought that this modification would take a time that I couldn't spend at that moment, but now I'm thinking about a way it can be done faster than I first thought.

Reply:
I agree that it should be split, having each key linked to a database key (instead of reloading the whole contract tree). However, in this moment, the most important part is to validate the format of the hash, which will never change in the future (even if we do this change to the persistance model).
Some working examples would be nice, so the community can easily understand the MPT process.

Reply:
> The entire MerklePatricia tree cannot be serialized and deserialized each time a block completes that changes storage of one of the items in the tree. The implementation needs to store the actual entries of the Merkle Patricia tree in separate keys in the LevelDB database, so that the entire tree is not written into level db for each contract that has had any storage change each block. This change is possible and required to be able to adopt the use of the MPT.

Hi @jsolman and @igormcoelho, I've used the MerklePatriciaDataCache class to wrap all the logic using some kind of delegates to make my approach uses the actual DataCache, by doing this it is not necessary to serialize/deserialize the entire MerklePatricia tree.
I've made this way because I wanted it to be easy to test as I did in the UT_MerklePatriciaDataCache, but I'm not very satisfied with one thing, which is that every time I need to set a pair of <MPTKey, MerklePatriciaNode> in the MPT I need to convert both to byte[] and the same thing when I need to get something from the MPT.
What do you think about it? Is it too much expensive?